include/io/mdarray.hpp

```diff
@@ -0,0 +1,274 @@
+        : m_shape(rank)
+        , m_data() {};
+
+    mdarray(ilist dsizes)
+        : m_shape(dsizes.begin(), dsizes.end())
+        , m_data(this->max_size()) {};
+
+    /**
+     * Retrieve a reference to the value at the given index.
+     *
+     * @example
+     * mdarray<int> x(2); // rank 2
+     * x.at({0, 0});      // get the value at (0, 0) in a 2D array.
+     * x.at({1, 1}) = 3;  // set the value at (1, 1) in a 2D array.
+     *
+     * @param n Index list
+     * @return reference 
+     */
+    reference at(const ilist& n)
+    {
+        return this->m_data.at(this->index(n));
+    }
+
+    /**
+     * Retrieve const reference to the value at the given index.
+     *
+     * @example
+     * mdarray<int> x(2); // rank 2
+     * x.at({0, 0});      // get the value at (0, 0) in a 2D array.
+     *
+     * @param n Index list
+     * @return reference 
+     */
+    const_reference at(const ilist& n) const
+    {
+        return this->m_data.at(this->index(n));
+    }
+
+    void insert(const ilist& n, value_type value)
+    {
+        for (size_type i = 0; i < this->m_shape.size(); i++) {
+            if (n[i] >= this->m_shape[i])
+                throw std::out_of_range("out of range");
+        }
+
+        size_type index = this->index(n);
+    
+        if (index > this->max_size())
+            throw std::out_of_range("out of range");
+        
+        if (index > this->size() || this->size() == 0)
+            this->m_data.resize(index + 1);
+    
+        this->m_data.at(index) = value;
+    }
+
+    void insert(std::initializer_list<std::pair<std::initializer_list<size_type>, value_type>> args)
+    {
+        for (const auto& arg : args) {
+            this->insert(arg.first, arg.second);
+        }
+    }
+
+    /**
+     * Allocate this mdarray based on the indices given.
+     * 
+     * @example
+     * mdarray<double> x(3);   // rank 3
+     * x.allocate({5, 10, 10}) // Allocates 5 * 10 * 10 elements
+     * @param n 
+     */
+    void allocate(const ilist& n)
+    {
+        this->m_data.resize(this->index(n));
+    }
+
+    /**
+     * Get the (total) size of this mdarray
+     * (aka the total number of elements).
+     * 
+     * @return size_type
+     */
+    size_type size() const noexcept
+    {
+        return this->m_data.size();
+    }
+
+    size_type max_size() const noexcept
+    {
+        size_type max = 1;
+        for (const auto& v : this->m_shape) {
+            max *= v;
+        }
+        return max;
+    }
+
+    iterator begin() const noexcept {
+        return iterator(*this, 0);
+    }
+
+    iterator end() const noexcept {
+        return iterator(*this, this->m_data.size());
+    }
+
+    /**
+     * Get the rank of this mdarray.
+     * 
+     * @return size_type 
+     */
+    size_type rank() const noexcept
+    {
+        return this->m_shape.size();
+    }
+
+    /**
+     * Get the shape of this mdarray.
+     * 
+     * @return ilist
+     */
+    ilist shape() const noexcept
```

Comment:
This is also going to force allocation of a short-lived `vector` on the callers. Ideally, we'd return `span<const size_t>`. I doubt callers would have life-time/aliasing issues with that in a case where `m_shape` changes.

Reply:
Or, the caller could pass in a `span` that they want filled in.

Reply:
Ah, I think this is something I forgot to fix. I realize I can return a reference, since `shape` is a member variable also I think.